man/man8/smtpd.8
.TH SMTPD 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/smtpd.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: smtpd.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
smtpd \- Smtp protocol server front end process
.SH
.in +.5i
.ti -.5i
/usr/lib/pp/smtpd
\%[-p\ port] \%[-i\ maxcon] server channel
.in -.5i
(under rc.local or inetd)
.SH DESCRIPTION
.I Smtpd
is the front end to the smtp server provided with PP. 
It provides a simple network listener service for smtp requests
similar to that provided by \fIinetd(8)\fP. However, the smtp server
it invokes can normally be run directly under inetd.
.PP
.I Smtpd
responds to the following flags.
.TP
\-p port
When running standalone, this overrides the default internet port that
\fIsmtpd\fP listens on.
.TP
\-i maxcon
When running standalone, this option limits the number of simultaneous
smtp connections that can be open. Once this limit is reached,
\fIsmtpd\fP will ignore further requests for connections until one of
the current connections terminates.
.PP
The two mandatory arguments to \fIsmtpd\fP are firstly the pathname of
the server that is to be executed to obey the smtp protocol, and
secondly the PP channel that the server should claim to originate from.
Neither of these values default and both must be specified.
.SH FILES
/etc/rc.local	started from here
.SH "SEE ALSO"
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP



CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Chans/smtp/smtpd.c
/* smtpd: daemon for smtp */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Chans/smtp/RCS/smtpd.c,v 6.0 1991/12/18 20:12:19 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Chans/smtp/RCS/smtpd.c,v 6.0 1991/12/18 20:12:19 jpo Rel $
 *
 * $Log: smtpd.c,v $
 * Revision 6.0  1991/12/18  20:12:19  jpo
 * Release 6.0
 *
 */



/*
 * new version - detects if it is running under inetd.
 */


#include "util.h"
#include <signal.h>
#include "sys.file.h"
#ifdef  SYS5
#ifdef vfork
#undef vfork
#endif
#include <unistd.h>
#endif
#include <sys/signal.h>
#include <isode/internet.h>
#include <sys/wait.h>
#include <sys/ioctl.h>
#include <errno.h>
#ifdef SYS5
#include <sys/resource.h>
#endif

static char usage[] = "usage: smptd [-p port] [-i maxconn] \
program channel";

static int smtpport;
static int	maxconn = 5;
static struct sockaddr_in rmtaddr, laddr;
static int	numconnections = 0;
static char thishost[LINESIZE];
static char *Channel, Smtpserver[BUFSIZ];
static char *nstimeout;

static void envinit ();
static SFD reaper ();

#if sparc && defined(__GNUC__)	/* work around bug in gcc 1.37 sparc version */
#define inet_ntoa myinet_ntoa

static char *myinet_ntoa (in)
struct in_addr in;
{
	static char buf[80];

	(void) sprintf (buf, "%d.%d.%d.%d",
			(in.s_addr >> 24) & 0xff,
			(in.s_addr >> 16) & 0xff,
			(in.s_addr >> 8 ) & 0xff,
			(in.s_addr	) & 0xff);
	return buf;
}
#else
extern char	*inet_ntoa ();
#endif

extern int errno;

char	*myname;
int	debug;

main (argc, argv)
int	argc;
char	**argv;
{
	extern char	*optarg;
	extern int	optind;
	int	opt;
	struct servent *sp;
	int	alen, i, skt;

	myname = argv[0];
	sys_init (myname);

	if ((sp = getservbyname ("smtp", "tcp")) == NULL)
		smtpport = htons (25);
	else	smtpport = sp -> s_port;

	while((opt = getopt(argc, argv, "p:i:t:")) != EOF)
		switch (opt) {
		    case 'i':
			maxconn = atoi (optarg);
			break;
		    case 'p':
			smtpport = htons ((short)atoi (optarg));
			break;
		    case 't':
			nstimeout = optarg;
			break;
		    default:
			fprintf (stderr, "bad switch -%c: %s\n", opt, usage);
			PP_OPER (NULLCP, ("Bad argument -%c: %s", opt, usage));
			exit (1);
		}
	argc -= optind;
	argv += optind;

	initialise (argc, argv);

	alen = sizeof rmtaddr;
	if (getpeername (0, (struct sockaddr *)&rmtaddr,
			 &alen) == 0) { /* under inetd */
		PP_TRACE (("Running under inetd..."));
		doit (0);
		_exit (0);
	}

	envinit ();

	laddr.sin_family = AF_INET;
	laddr.sin_addr.s_addr = INADDR_ANY;
	laddr.sin_port = smtpport;

	for (i = 0; i < 10; sleep ((unsigned) 2 * ++i)) {
		if ((skt = socket (AF_INET, SOCK_STREAM, 0)) == -1) {
			PP_SLOG (LLOG_EXCEPTIONS, "socket",
				 ("can't create "));
			continue;
		}

		if (bind (skt, (struct sockaddr *)&laddr,
			  sizeof laddr) == -1) {
			PP_SLOG (LLOG_EXCEPTIONS, "bind",
				 ("socket can't "));
			(void) close (skt);
			continue;
		}

		if (listen (skt, 5) == -1) {
			PP_SLOG (LLOG_EXCEPTIONS, "listen",
				 ("socket can't "));
			(void) close (skt);
			continue;
		}
		break;
	}

	PP_NOTICE (("%s listening on port %d", myname, ntohs(smtpport)));
#define tcp_still_alive_and_kicking 1

	while (tcp_still_alive_and_kicking) {
		int	snew;
		int	pid;

		alen = sizeof (rmtaddr);
		if ((snew = accept (skt, (struct sockaddr *)&rmtaddr,
				    &alen)) == NOTOK) {
			if (errno != EINTR)
				PP_SLOG (LLOG_EXCEPTIONS, "failed",
					 ("accept "));
			continue;
		}
		PP_TRACE (("Accepted new connection from %s",
			   inet_ntoa (rmtaddr.sin_addr)));
		if ((pid = tryfork ()) == NOTOK) {
			PP_SLOG (LLOG_EXCEPTIONS, "failed", ("fork "));
			sleep (10);
			continue;
		}
		if (pid == 0) {	/* child */
			doit (snew);
			_exit(0);
		}
		(void) close (snew);

		numconnections ++;

		while (numconnections >= maxconn) {
			PP_TRACE (("Too many connections -- waiting"));
#ifdef SYS5
			reaper(0);
#else
			sigpause (0);
#endif
			PP_TRACE (("Finished waiting"));
		}
	}
}

/* ARGSUSED */
static SFD reaper (sig)
int sig;
{
#ifndef WNOHANG
	int status;
	alarm(2);
	while (wait(&status) != -1)
		numconnections --;
	alarm (0);
#else
#ifdef SYS5
	int status;
#else
	union wait status;
#endif
	while (wait3 (&status, WNOHANG, (caddr_t)0) > 0)
		numconnections --;
#endif
}

#ifdef SYS5
SFD alarmtrap (sig)
int sig;
{
	(void) signal (SIGALRM, alarmtrap);
}
#endif

doit (fd)
int	fd;
{
	char *argv[10];
	int argc = 0;

	if (fd != 0)
		(void) dup2 (fd, 0);
	if (fd != 1)
		(void) dup2 (fd, 1 );
	if (fd != 2)
		(void) dup2 (fd, 2 );
	if (fd > 2 )
		(void) close (fd);
	argv[argc++] = Channel; /* argv[0] */
	if (nstimeout) {
		argv[argc++] = "-t";
		argv[argc++] = nstimeout;
	}
	argv[argc++] = Channel;
	argv[argc] = NULLCP;
	execv (Smtpserver, argv);
	PP_OPER (Smtpserver, ("server exec "));
	_exit (99);
}

static void  envinit () 
{
	int     i,
		nbits,
		sd;

	nbits = getdtablesize ();

	if (!(debug = isatty (2))) {
		for (i = 0; i < 5; i++) {
			switch (fork ()) {
			    case NOTOK: 
				sleep (5);
				continue;

			    case OK: 
				break;

			    default: 
				_exit (0);
			}
			break;
		}

		if ((sd = open ("/dev/null", O_RDWR)) == NOTOK) {
			PP_OPER ("/dev/null", ("unable to read"));
			exit (3);
		}
		if (sd != 0)
			(void) dup2 (sd, 0), (void) close (sd);
		(void) dup2 (0, 1);
		(void) dup2 (0, 2);

#ifdef  TIOCNOTTY
		if ((sd = open ("/dev/tty", O_RDWR)) != NOTOK) {
			(void) ioctl (sd, TIOCNOTTY, NULLCP);
			(void) close (sd);
		}
#else
#ifdef  SYS5
		(void) setpgrp ();
		(void) signal (SIGINT, SIG_IGN);
		(void) signal (SIGQUIT, SIG_IGN);
#endif
#endif
	}
	else
		ll_dbinit (pp_log_norm, myname);

#ifdef notdef			/* damn YP... */
	for (sd = 3; sd < nbits; sd++)
		(void) close (sd);
#endif

	(void) signal (SIGPIPE, SIG_IGN);

	ll_hdinit (pp_log_norm, myname);
#ifndef SYS5
	(void) signal (SIGCHLD, reaper);
#endif

	PP_TRACE (("starting"));
}

initialise (argc, argv)
int	argc;
char	**argv;
{
	int	i;
	struct hostent *hp;
	extern char *chndfldir;

	if (argc > 0)
	{
		getfpath (chndfldir, argv[0], Smtpserver);

		if (access (Smtpserver, X_OK) < 0 ) { /* execute privs? */
			PP_OPER (Smtpserver, ("Cannot access server program"));
			fprintf (stderr, "Cannot access server program %s\n",
				 Smtpserver);
			exit (99);
		}
		PP_TRACE (("server is '%s'", Smtpserver ));
	} else {
		fprintf (stderr, "Smtpserver program not specified! (%s)\n",
			 usage);
		PP_OPER (NULLCP, ("Smtpserver program not specified! (%s)",
				  usage));
		exit (99);
	}

	if( argc > 1 ) {
		Channel = argv[1];
		PP_TRACE (("channel is '%s'", Channel));
	} else {
		PP_OPER (NULLCP, ("Channel not specified! (%s)", usage));
		fprintf (stderr, "Channel not specified! (%s)\n", usage);
		exit (99);
	}

        /*
         * must get full name for this host from nameserver (or file)
         */

	if (gethostname (thishost, sizeof thishost) == -1)
		PP_SLOG (LLOG_EXCEPTIONS, "gethostname",
			 ("Can't find out who I am"));

        for(i = 0 ; i < 10 ; i++){
                hp = gethostbyname(thishost);
                if(hp != NULL)
                        break;
                if(i > 2)
                        sleep((unsigned)i*2);
        }
        if(hp == NULL) {
		PP_OPER (NULLCP, 
                  ("Cannot find 'official' name of host '%s'\n",
						thishost));
                fprintf(stderr,
		  "Cannot find 'official' name of host '%s'\n",
                                                   thishost);
                exit(-1);
        }
        (void) strcpy(thishost, hp->h_name);
#ifdef SYS5
	(void) signal (SIGALRM, alarmtrap);
#endif
}
